In [1]:
import cv2
import numpy as np

# --------------------------------------------------------------------------------------------------------------

# Reading the Images:-

###    To read images in OpenCV we use `cv2.imread()` Method 
#### cv2.imread:-[LINK](https://docs.opencv.org/4.x/d4/da8/group__imgcodecs.html#ga288b8b3da0892bd651fce07b3bbd3a56)
#### IMREAD FLAGS:[LINK](https://docs.opencv.org/4.x/d8/d6a/group__imgcodecs__flags.html#ga61d9b0126a3e57d9277ac48327799c80)


In [2]:
img3 = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\graf1.png")
print(img3)

[[[215 209 219]
  [215 206 216]
  [207 196 206]
  ...
  [ 27  29  20]
  [ 28  29  31]
  [ 18  17  30]]

 [[211 206 217]
  [217 210 221]
  [219 209 220]
  ...
  [ 42  33  29]
  [ 35  24  28]
  [ 32  17  32]]

 [[213 210 222]
  [215 209 222]
  [215 206 219]
  ...
  [ 39  29  30]
  [ 39  27  32]
  [ 35  22  35]]

 ...

 [[ 29  23 178]
  [ 22  20 174]
  [ 22  25 176]
  ...
  [ 47  37  39]
  [ 46  36  37]
  [ 51  40  40]]

 [[ 17  23 178]
  [ 24  29 186]
  [ 28  32 189]
  ...
  [ 47  41  41]
  [ 47  39  41]
  [ 47  40  41]]

 [[ 23  33 185]
  [ 26  31 186]
  [ 30  29 188]
  ...
  [ 43  44  42]
  [ 40  43  43]
  [ 35  37  41]]]


In [3]:
img2 = cv2.imread(r"dssdas.png")
print(img2)

None


### We can see in the above cells that Images are represented using NUMPY Arrays with a dtype of `unit8`(return    of `imread` Function)

#### 1)Each cell in the Array represents a pixel having values ranging from `0-255`
#### 2)If we provide a Non-exsisting file to `imread` method then it will not throw an error but will return None Object

In [4]:
img3.shape

(640, 800, 3)

### Width:-800 and Height:-640 of the image

## But as we can see above that in The NUMPY Array its represented as (Height,width,Channels)

### This above phenomena is bcoz  of how the numpy arrays as represented as (Rows,Columns,...)
### Number of Rows = Height of the Image
### Number of Columns = Width of the Image

In [5]:
img4 = cv2.imread(r"E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\InputData\graf1.png",cv2.IMREAD_GRAYSCALE)
print(img4)

[[212 210 200 ...  26  29  21]
 [209 214 213 ...  32  26  23]
 [213 213 210 ...  30  29  27]
 ...
 [ 70  66  69 ...  38  37  41]
 [ 68  75  78 ...  41  40  41]
 [ 77  76  76 ...  43  42  37]]


In [6]:
img4.shape

(640, 800)

### As we can in above cells that we read the image as GRAYSCALE image which only have one channel so the equalvalent NUMPY array is represented as (Hieght,Width)

# --------------------------------------------------------------------------------------------------------------

# Displaying The Images:-

## To display the Image represented by an NUMPY Array we use:-
### 1)`cv2.imshow()` Method :-It recieves < Window Name > , < Numpy Array > as main arguments  and it displays  the Image represented by array on the window of the same name,If the Window doesnot exsists it creates it(Default Settings Window Created)
### 2)`cv2.waitKey()` Method :-It recieves < Integer(n) > as main arguments and it pauses the script for n seconds.  If n was not passed or equal to 0 then it hold for infinite time.It only continues the script(for both cases) If it recives a input from keyboard or the Image Window was closed.     

### 3)`cv2.destroyAllWindows()` Method:-It destroys all the Window Images that were openned

#### cv2.imread():-[LINK](https://docs.opencv.org/4.x/d7/dfc/group__highgui.html#ga453d42fe4cb60e5723281a89973ee563)
#### cv2.waitKey():-[LINK](https://docs.opencv.org/4.x/d7/dfc/group__highgui.html#ga453d42fe4cb60e5723281a89973ee563)

In [7]:
cv2.imshow("Image",img3)
cv2.waitKey(0)
cv2.destroyAllWindows()

### As Running in the Cell Above we can see that imshow open an Window(with default settings) to show the image and all the respective features described in description

In [8]:
cv2.namedWindow("Image1Window",cv2.WINDOW_NORMAL)
cv2.imshow("Image1Window",img3)
cv2.waitKey()
cv2.destroyAllWindows()

### The `cv2.namedWindow()` Method is used to create a WIndow by the User with specfic fetures which can be decided by the Flags send in the agruments

#### cv2.namedWindow():- [LINK](https://docs.opencv.org/4.x/d7/dfc/group__highgui.html#ga5afdf8410934fd099df85c75b2e0888b)
#### NAMEDWINDOW Flags():- [LINK](https://docs.opencv.org/4.x/d0/d90/group__highgui__window__flags.html#ggabf7d2c5625bc59ac130287f925557ac3acf621ace7a54954cbac01df27e47228f)

# -------------------------------------------------------------------------------------------------------------

# Writing The Image:-

## To write an Image represented in NUMPY Array form to the disk we use `cv2.imwrite()` Method

#### cv2.imwrite():- [LINK](https://docs.opencv.org/4.x/d4/da8/group__imgcodecs.html#gabbc7ef1aa2edfaa87772f1202d67e0ce)

In [9]:
cv2.imwrite("E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\OutputData\OpenCV101\\abc.jpg",img3)

True

### As we can see in the above cell we passed < Absolute Image Location with the filename and extension to be saved as > as first agrument and < NUMPY Array respresenting the Image" as second argument

### NOTE:-It will automactically use Defualt Flag while writing the image according to its extension.

In [10]:
cv2.imwrite("E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\OutputData\OpenCV101\\abc1.jpeg",img3,[cv2.IMWRITE_JPEG_QUALITY,10])
cv2.imwrite("E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\OutputData\OpenCV101\\abc2.jpeg",img3,[cv2.IMWRITE_JPEG_QUALITY,100])
cv2.imwrite("E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\OutputData\OpenCV101\\abc3.jpeg",img3,[cv2.IMWRITE_JPEG_LUMA_QUALITY,100])
cv2.imwrite("E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\OutputData\OpenCV101\\abc4.png",img3,[cv2.IMWRITE_PNG_BILEVEL,0])
cv2.imwrite("E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\OutputData\OpenCV101\\abc5.ppm",img3,[cv2.IMWRITE_PXM_BINARY,0])
cv2.imwrite("E:\PyImage_ComputerVision\OpenCVBasic\WorkingData\OutputData\OpenCV101\\abc6.ppm",img3,[cv2.IMWRITE_PXM_BINARY,1])

True

### In the Above Cell we can see we have used different flags to store the Image with different set of features etc.
# ***BeCareful to use same flags with same Extensions

## We can CONVERT IMAGE from one extension to another by reading the Image in the original extension and then writing in the extension we want

# ---------------------------------------------------------------------------------------------------------------